### PR TITLE
input validation added to roc

### LINKED ISSUE #98: Multiclass ROC curve
ROC curve is commonly used to compare the performance of models. It is usually used in binary classification, but it can also be used in multiclass classification using averaging methods.

When running `plot.roc(y_test, prediction1)` on more than 2 classes it fails, we should add support to it:

`ValueError: multiclass format is not supported`

I tried it for ROC curve, but this issue also applies to PR curve. (`plot.precision_recall(y_test, prediction1`)

### PR DESCRIPTION
**Pull request by yafimvo**
## Describe your changes
1. Input validation added to roc
2. Merged with the latest version (issue 146)
3. Removed the ROC testing image base to `test_roc`

`plot.ROC.from_raw_data` and `plot.roc` take 2 inputs: `y_test` and `y_score` in the following formats

### "classes"

```
[0, 1, 2, 0, 1, ...]
```

```
['virginica', 'versicolor', 'virginica', 'setosa', ...]
```

### "one-hot encoded classes"

```
[[0, 0, 1],
 [1, 0, 0]]
```

### "scores"
```
[[0.1, 0.1, 0.8],
 [0.7, 0.15, 0.15]]
```

## Issue ticket number and link
Closes #98 

## Checklist before requesting a review
- [X] I have performed a self-review of my code
- [X] I have added thorough tests (when necessary).
- [X] I have added the right documentation (when needed). Product update? If yes, write one line about this update.

----
:books: Documentation preview :books:: https://sklearn-evaluation--230.org.readthedocs.build/en/230/

### CODE REVIEW

**Comment by idomic** [comment]
@yafimvo please resolve conflicts

**Review by edublancas — CHANGES_REQUESTED** [review]
found two issues:


image

gives this error:

```
ValueError: classes [0 1] mismatch with the labels [0 1 2] found in the data
```

looks like it thinks it only has  two classes, but it has three. but where is the 2 coming from?


to reproduce:

```python
import numpy as np
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn_evaluation import plot
from sklearn import preprocessing

iris = load_iris()
X, y = iris.data, iris.target
y_ = iris.target_names[y]

random_state = np.random.RandomState(0)
n_samples, n_features = X.shape

X = np.concatenate([X, random_state.randn(n_samples, 200 * n_features)], axis=1)
(
    X_train,
    X_test,
    y_train,
    y_test,
) = train_test_split(X, y_, test_size=0.5, stratify=y, random_state=0)

classifier = LogisticRegression()
y_score = classifier.fit(X_train, y_train).predict_proba(X_test)

lb = preprocessing.LabelBinarizer()
lb.fit(y_test)
y_test_bin = lb.transform(y_test)

roc = plot.ROC.from_raw_data(y_test_bin, y_score)
```


also, when calling this (same code as above):

```
# try to break it by passing y_test_bin in y_score
roc = plot.ROC.from_raw_data(y_test, y_test_bin)
````

the error shows:

```
ValueError: Please check y_score values. 
Expected scores array-like. got: [[0 0 1]
 [0 1 0]
 [0 0 1]
 [1 0 0]
 [0 0 1]
 [1 0 0]
 [0 1 0]
 [1 0 0]
 [0 0 1]
 [0 1 0]
 [1 0 0]
[really long error]
```

i see that the error is displaying the whole input, however, it's very long. It'd be better to just display the first few characters:

probably something like this:

image

**Comment by yafimvo** [comment]
@edublancas 

[One of our examples](https://sklearn-evaluation.ploomber.io/en/latest/classification/compare.html#roc) is trying to plot a roc curve using the `DecisionTreeClassifier` and `y_score` is in an invalid format

```
tree_score, forest_score = [
    est.fit(X_train, y_train).predict_proba(X_test)
    for est in [DecisionTreeClassifier(), RandomForestClassifier()]
]
```

Do we need to support it?

If not, we can change it to `LogisticRegression` and it will work fine.

File: src/sklearn_evaluation/plot/roc.py [hunk 1]
```diff
@@ -1,16 +1,135 @@
...
+    is_numeric = np.issubdtype(array.dtype, np.number)
+
+    if is_numeric:
+        array_flatten = array.flatten()
+
+        # check if array consists of 0 and 1
+        _is_binary = is_binary(array)
+
+        if _is_binary:
+            _is_1d_array = len(array.shape) == 1
```
Comment (neelasha23) [inline src/sklearn_evaluation/plot/roc.py:116]: can we use switch case here for better readability?

File: src/sklearn_evaluation/plot/roc.py [hunk 2]
```diff
@@ -349,3 +484,39 @@ def _calculate_plotting_data(y_true, y_score):
             tpr = [tpr]
 
         return fpr, tpr, label
+
+
+def _preprocess_array_for_roc(array):
+    """
+    https://github.com/ploomber/sklearn-evaluation/pull/221/files
```
Comment (neelasha23) [inline src/sklearn_evaluation/plot/roc.py:491]: is this comment needed here?

File: src/sklearn_evaluation/util.py [hunk 3]
```diff
@@ -39,14 +40,37 @@ def _can_iterate(obj):
     return is_iterable and not is_string
 
 
+def check_elements_in_range(array, min, max, include_min=True, inclue_max=True):
```
Comment (neelasha23) [inline src/sklearn_evaluation/util.py:43]: It can be renamed to `get_elements_in_range`

File: tests/test_roc.py [hunk 4]
```diff
@@ -161,3 +175,199 @@ def test_fpr_and_tpr_different_types(fpr, tpr):
...
+):
+    y_test, y_score = roc_multi_classification_raw_data
+    y_test_roc2, y_score_roc2 = roc_multi_classification_raw_data_set2
+
+    roc1 = plot.ROC.from_raw_data(y_test, y_score)
+    roc2 = plot.ROC.from_raw_data(y_test_roc2, y_score_roc2)
+    roc1 + roc2
+
+
+@image_comparison(baseline_images=["roc_multi_set2", "roc_multi_add_multi"])
```
Comment (neelasha23) [inline tests/test_roc.py:364]: One more image can be added to the list if the plot is generating in the test (roc1, roc2, roc1+roc2)

File: src/sklearn_evaluation/util.py (hunk 3 above)
Comment (yafimvo) [reply src/sklearn_evaluation/util.py:43]: I think `get_elements_in_range` implies a different case.
This method checks if values in an array are within a range. I can change it to something like `check_array_range`. what do you think? @neelasha23

File: src/sklearn_evaluation/plot/roc.py (hunk 2 above)
Comment (yafimvo) [reply src/sklearn_evaluation/plot/roc.py:491]: This method is not in use and @edublancas asked me to keep it so I wanted to refer to the original discussion. I'll remove it and add a better description.

File: src/sklearn_evaluation/util.py (hunk 3 above)
Comment (neelasha23) [reply src/sklearn_evaluation/util.py:43]: ok I got it now!

File: src/sklearn_evaluation/plot/roc.py (hunk 1 above)
Comment (neelasha23) [reply src/sklearn_evaluation/plot/roc.py]: yes that's fine I guess. Maybe some short comments can also help.

**Comment by edublancas** [comment]
> In [one of our examples](https://sklearn-evaluation.ploomber.io/en/latest/classification/compare.html#roc) we try to plot a roc curve using the DecisionTreeClassifier and as a result, y_score is in an invalid format

good catch. technically speaking, this is in the right format since these are [valid scores](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html#sklearn.tree.DecisionTreeClassifier.predict_proba) (they are floats, and 0.0. and 1.0 are valid values). the "issue" here is the nature of the decision tree model, which tends to produce these types of scores.

I think let's change it to logistic regression so the scores are more meaningful

**Comment by idomic** [comment]
> I think let's change it to logistic regression so the scores are more meaningful

In this specific case for roc or the whole guide?

@edublancas @yafimvo Also what else is missing here?

**Comment by yafimvo** [comment]
@idomic 
I changed the example only for roc.

Ready for review

File: tests/test_roc.py [hunk 5]
```diff
@@ -161,3 +175,201 @@ def test_fpr_and_tpr_different_types(fpr, tpr):
...
+    ).T
+
+    roc1 = plot.ROC.from_raw_data(y_test, y_score)
+    roc2 = plot.ROC.from_raw_data(y_test_roc2, y_score_roc2)
+    roc1 + roc2
+
+
+@image_comparison(baseline_images=["roc_add_roc"])
+def test_roc_add_to_roc(roc_values):
+    fpr1, tpr1 = roc_values
```
Comment (idomic) [inline tests/test_roc.py:318]: Why do we need this? it seems like those values get overwritten?
I see similar stuff in other test cases as well

Comment (idomic) [reply tests/test_roc.py:318]: If we can parameterize the inputs I think we should do it as the test seems pretty much the same for the roc add

**Review by idomic — CHANGES_REQUESTED** [review]
looks good, just added a comment on the tests

File: tests/test_roc.py (hunk 5 above)
Comment (yafimvo) [reply tests/test_roc.py:318]: Good point. This was from the previous code where roc tests were scattered in different places. I removed all roc resources from `conftest.py` to `test_roc.py` and parameterized the "short" values that are easy to use. The more complex values I kept as a fixture.

**Comment by idomic** [comment]
Nice job @yafimvo !

**Pull request merged**